File: web/blueprint/api_v1/routes/language.py
from werkzeug import Response

from web.api import API, json_response
from web.auth import authorize
from web.blueprint.api_v1 import api_v1_bp
from web.database import conn
from web.database.model import Language, UserRoleLevel

#
# Configuration
#


class LanguageAPI(API):
    model = Language
    post_columns = {
        Language.code,
        Language.in_sitemap,
        Language.name,
    }
    get_columns = {
        Language.code,
        Language.in_sitemap,
        Language.name,
    }


#
# Endpoints
#


@api_v1_bp.post("/languages")
@authorize(UserRoleLevel.ADMIN)
def post_languages() -> Response:
    api = LanguageAPI()
    data = api.gen_data(api.post_columns)
    with conn.begin() as s:
        model = api.model()
        api.insert(s, data, model)
        resource = api.gen_resource(s, model)
    return json_response(data=resource)


@api_v1_bp.get("/languages")
def get_languages() -> Response:
    api = LanguageAPI()
    with conn.begin() as s:
        models: list[Language] = api.list_(s)
        resources = api.gen_resources(s, models)
    return json_response(data=resources)


@api_v1_bp.get("/languages/<int:language_id>")
def get_languages_id(language_id: int) -> Response:
    api = LanguageAPI()
    with conn.begin() as s:
        model: Language = api.get(s, language_id)
        resource = api.gen_resource(s, model)
    return json_response(data=resource)


@api_v1_bp.delete("/languages/<int:language_id>")
@authorize(UserRoleLevel.ADMIN)
def delete_languages_id(language_id: int) -> Response:
    api = LanguageAPI()
    with conn.begin() as s:
        model: Language = api.get(s, language_id)
        api.delete(s, model)
    return json_response()


#
# Functions
#

File: web/blueprint/admin/static/js/api/language.js
function postLanguages(data, silent = false) {
    const url = `/api/v1/languages`;
    return callApi('POST', url, data, 'application/json', silent);
}

function getLanguages(silent = false) {
    const url = `/api/v1/languages`;
    return callApi('GET', url, null, null, silent);
}

function getLanguagesId(languageId, silent = false) {
    const url = `/api/v1/languages/${languageId}`;
    return callApi('GET', url, null, null, silent);
}

function deleteLanguagesId(languageId, silent = false) {
    const url = `/api/v1/languages/${languageId}`;
    return callApi('DELETE', url, null, null, silent);
}